### ISSUE TITLE
[rush] ERROR: EEXIST: file already exists, symlink linking dependencies

### ISSUE THREAD

**Issue by jbcpollak** [Microsoft/web-build-tools#788]
Environment: MacOS, Node 10.4.1, npm 6.4.0

The following command works fine with rush 5.0.0, but fails with 5.0.4:

`rush update -p --full --debug`

I get the following output:

```
npm verb saving []
npm verb shrinkwrap skipping write for package.json because there were no changes.
npm info lifecycle undefined~preshrinkwrap: undefined
npm info lifecycle undefined~shrinkwrap: undefined
npm notice created a lockfile as package-lock.json. You should commit this file.
npm info lifecycle undefined~postshrinkwrap: undefined
npm WARN ajv-keywords@3.2.0 requires a peer of ajv@^6.0.0 but none is installed. You must install peer dependencies yourself.

added 1914 packages from 2112 contributors and audited 64930 packages in 42.914s
found 61 vulnerabilities (33 low, 28 high)
  run `npm audit fix` to fix them, or `npm audit` for details
npm verb exit [ 0, true ]
npm timing npm Completed in 43238ms
npm info ok

Running "npm shrinkwrap"...
npm info it worked if it ends with ok
npm verb cli [ '/Users/jpollak/.nodenv/versions/10.4.1/bin/node',
npm verb cli   '/Users/jpollak/src/wis/common/temp/npm-local/node_modules/.bin/npm',
npm verb cli   'shrinkwrap',
npm verb cli   '--cache',
npm verb cli   '/Users/jpollak/src/wis/common/temp/npm-cache',
npm verb cli   '--tmp',
npm verb cli   '/Users/jpollak/src/wis/common/temp/npm-tmp',
npm verb cli   '--verbose' ]
npm info using npm@6.4.0
npm info using node@v10.4.1
npm notice package-lock.json has been renamed to npm-shrinkwrap.json. npm-shrinkwrap.json will be used for future installations.
npm verb exit [ 0, true ]
npm timing npm Completed in 220ms
npm info ok
"npm shrinkwrap" completed

Updating /Users/jpollak/src/wis/common/config/rush/npm-shrinkwrap.json

Linking projects together...

LINKING: @sixriver/wis-commit-hooks
Purging /Users/jpollak/src/wis/commit-hooks/node_modules

ERROR: EEXIST: file already exists, symlink '/Users/jpollak/src/wis/commit-hooks/node_modules/@commitlint/prompt-cli/CHANGELOG.md' ->
'/Users/jpollak/src/wis/common/temp/node_modules/@commitlint/prompt-cli/CHANGELOG.md'
```

**Comment by jbcpollak** [comment]
I confirmed this problem is present in 5.0.2 and 5.0.3. It appears 5.0.1 doesn't exist?

**Comment by pgonzal** [comment]
Right, 5.0.1 was skipped because of an issue encountered during publishing

**Comment by jbcpollak** [comment]
```
ERROR: EEXIST: file already exists, symlink '/Users/jpollak/src/wis/commit-hooks/node_modules/@commitlint/prompt-cli/CHANGELOG.md' ->
'/Users/jpollak/src/wis/common/temp/node_modules/@commitlint/prompt-cli/CHANGELOG.md'

Error: EEXIST: file already exists, symlink '/Users/jpollak/src/wis/commit-hooks/node_modules/@commitlint/prompt-cli/CHANGELOG.md' -> '/Users/jpollak/src/wis/common/temp/node_modules/@commitlint/prompt-cli/CHANGELOG.md'
    at Object.fs.symlinkSync (fs.js:1002:3)
    at Function.createSymbolicLinkFile (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/node-core-library/lib/FileSystem.js:263:13)
    at Function._createSymlink (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/base/BaseLinkManager.js:32:48)
    at Function._createSymlinksForDependencies (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/base/BaseLinkManager.js:107:37)
    at Function._createSymlinksForTopLevelProject (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/base/BaseLinkManager.js:54:33)
    at NpmLinkManager._linkProject (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/npm/NpmLinkManager.js:222:24)
    at _readPackageTree.then (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/npm/NpmLinkManager.js:26:22)
```

**Comment by pgonzal** [comment]
I reviewed PR #727 (which is the most related-seeming recent code change since Rush 5.0.0).  It replaced this code:

```ts
  protected static _createSymlink(linkTarget: string, linkSource: string, symlinkKind: SymlinkKind): void {
    fsx.mkdirsSync(path.dirname(linkSource));

    if (symlinkKind === SymlinkKind.Directory) {
      // For directories, we use a Windows "junction".  On Unix, this produces a regular symlink.
      fsx.symlinkSync(linkTarget, linkSource, 'junction');
    } else {
      if (process.platform === 'win32') {
        // For files, we use a Windows "hard link", because creating a symbolic link requires
        // administrator permission.
        fsx.linkSync(linkTarget, linkSource);
      } else {
        // However hard links seem to cause build failures on Mac, so for all other operating systems
        // we use symbolic links for this case.
        fsx.symlinkSync(linkTarget, linkSource, 'file');
      }
    }
  }
```

...with this code:
```ts
  protected static _createSymlink(linkTarget: string, linkSource: string, symlinkKind: SymlinkKind): void {
    FileSystem.ensureFolder(path.dirname(linkSource));

    if (symlinkKind === SymlinkKind.Directory) {
      // For directories, we use a Windows "junction".  On Unix, this produces a regular symlink.
      FileSystem.createSymbolicLinkJunction(linkTarget, linkSource);
    } else {
      if (process.platform === 'win32') {
        // For files, we use a Windows "hard link", because creating a symbolic link requires
        // administrator permission.
        FileSystem.createHardLink(linkTarget, linkSource);
      } else {
        // However hard links seem to cause build failures on Mac, so for all other operating systems
        // we use symbolic links for this case.
        FileSystem.createSymbolicLinkFile(linkTarget, linkSource);
      }
    }
  }
```

This adds the call to createSymbolicLinkFile() seen in your callstack:

```
    at Object.fs.symlinkSync (fs.js:1002:3)
    at Function.createSymbolicLinkFile (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/node-core-library/lib/FileSystem.js:263:13)
    at Function._createSymlink (/Users/jpollak/.nodenv/versions/10.4.1/lib/node_modules/@microsoft/rush/node_modules/@microsoft/rush-lib/lib/logic/base/BaseLinkManager.js:32:48)
    at Function._createSymlinksForDependencies 
```

The new API does this:
```ts
  public static createSymbolicLinkFile(linkTarget: string, linkSource: string): void {
    fsx.symlinkSync(linkSource, linkTarget, 'file');
  }
```

The parameters are reversed.  I'll create a PR to fix it.

@nickpape-msft  FYI

**Comment by pgonzal** [comment]
Thanks @jbcpollak for taking the time to report this!

**Comment by pgonzal** [comment]
@jbcpollak   This fix was published as Rush 5.0.5.  (Let us know if not.)  Thanks again!
